transactions/src/main/java/org/jboss/as/txn/service/internal/tsr/JCAOrderedLastSynchronizationList.java

```diff
@@ -87,7 +107,18 @@ public void addInterposedSynchronization(Synchronization synchronization) {
      */
     @Override
     public void beforeCompletion() {
-        for (Synchronization preJcaSync : preJcaSyncs) {
+        // This is needed to guard against syncs being registered during the run
+        int lastIndexProcessed = -1;
+        Synchronization[] copiedPreJcaSyncs = (Synchronization[]) preJcaSyncs.toArray(new Synchronization[] {});
```

Comment:
I don't understand why this copy is necessary? Couldn't you just do an index (rather than iterator) based iteration of preJcaSync?